src/MagicWormhole/Internal/Sequential.hs

```diff
@@ -0,0 +1,64 @@
+-- | Helpers for processing messages in sequence.
+module MagicWormhole.Internal.Sequential
+  ( Sequential
+  , sequenceBy
+  , insert
+  , next
+  ) where
+
+import Protolude
+
+import Control.Concurrent.STM.TVar
+  ( TVar
+  , modifyTVar'
+  , newTVar
+  , readTVar
+  )
+import qualified Data.HashMap.Lazy as HashMap
+
+-- | A thing that gets numbered items and then buffers them so you can read
+-- them in order without gaps.
+--
+-- "counter" is an 'Enum' that is used to determine the ordering of the
+-- elements, and "item" is the type of the items themselves.
+data Sequential counter item
+  = Sequential
+  { -- | The current value of the counter. The next item must have this rank.
+    current :: TVar counter
+  , -- | Where we store the items that have arrived out of order. A priority
+    -- queue might be a better choice.
+    buffer :: TVar (HashMap.HashMap counter item)
+  , -- | How to rank items. Each item has a "rank", which is an 'Enum'
+    -- (specifically, a "counter"). Items are yielded in order, without gaps,
+    -- by 'next'.
+    rank :: item -> counter
+  }
+
+-- | Create a 'Sequential' value for a series of item.
+sequenceBy
+  :: (Hashable counter, Eq counter)
+  => (a -> counter) -- ^ How to rank items
+  -> counter -- ^ The expected rank of the first item
+  -> STM (Sequential counter a)
+sequenceBy rank' initial = Sequential <$> newTVar initial <*> newTVar mempty <*> pure rank'
+
+-- | Insert an item into the sequence. It may only be drawn from the
+-- sequence with 'next' when its counter is next one.
+insert :: (Enum counter, Hashable counter, Eq counter) => Sequential counter a -> a -> STM ()
+insert sequential msg = modifyTVar' (buffer sequential) (HashMap.insert (rank sequential msg) msg)
```

Comment:
If `current` is greater than `sequential` then this is a bad insertion to make.  Worth catching somehow?

Reply:
Yes. I'll change this to raise an exception / return an error.

Reply:
Changed to return a `Bool` that indicates insertion success.